rtrlib/aspa/aspa.c

```diff
@@ -0,0 +1,267 @@
+/*
+ * This file is part of RTRlib.
+ *
+ * This file is subject to the terms and conditions of the MIT license.
+ * See the file LICENSE in the top level directory for more details.
+ *
+ * Website: http://rtrlib.realmv6.org/
+ */
+
+#include "aspa_private.h"
+#include "ordered_dyn_array/ordered_dyn_array.h"
+#include "rtrlib/rtrlib_export_private.h"
+#include "rtrlib/lib/alloc_utils_private.h"
+
+#include <pthread.h>
+#include <stdio.h>
+#include <string.h>
+#include <assert.h>
+
+static void aspa_table_notify_clients(struct aspa_table *aspa_table, const struct aspa_record *record, const struct rtr_socket *rtr_socket, const bool added)
+{
+	if (aspa_table->update_fp) {
+		// Realloc in order not to expose internal record
+		struct aspa_record *rec = lrtr_realloc((void *)record, sizeof(record) + record->provider_count * sizeof(record->provider_asns[0]));
+		aspa_table->update_fp(aspa_table, *rec, rtr_socket, added);
+	}
+}
+
+RTRLIB_EXPORT void aspa_table_init(struct aspa_table *aspa_table, aspa_update_fp update_fp)
+{
+	aspa_table->update_fp = update_fp;
```

Comment:
is it desired to have the aspa_table->cache value be undetermined?

Reply:
no, you're right